### PR TITLE
Experimental Swarm Implementation (2nd Storage Network Support)

### PR DESCRIPTION
**Pull request by bonedaddy**
## :construction_worker: Purpose

This enables basic support for a 2nd storage network that can be accessed by Temporal. IT allows basic file uploads to the Swarm storage network, sometimes called "Ethereum Swarm" or "etherswarm". 

We use a basic wrapper [client](https://github.com/RTradeLtd/swampi)  to abstract low-level usage of the swarm network. It also allows high-level data replication by distributing data to two different swarm nodes.

## :rocket: Changes

Allow swarm uploads

## :warning: Breaking Changes

None

### CODE REVIEW

File: api/v2/routes_rtfsp_management.go [hunk 1]
```diff
@@ -30,9 +30,11 @@ func (api *API) createIPFSNetwork(c *gin.Context) {
...
 		return
 	}
-	// make sure the name is something other than public
-	if strings.ToLower(networkName) == "public" {
-		Fail(c, errors.New("network name can't be public, or PUBLIC"))
+	switch strings.ToLower(networkName) {
+	case "public", "etherswarm":
+		Fail(c, errors.New("networok name cant be public, PUBLIC, etherswarm, or ETHERSWARM"))
+		return
+	default:
```
Comment (xiegeo) [inline api/v2/routes_rtfsp_management.go:37]: I think this default line is unnecessary.

File: api/v2/api.go [hunk 2]
```diff
@@ -62,7 +63,10 @@ type API struct {
 	queues      queues
 	service     string
 	version     string
-
+	swarm       struct {
+		endpoint1 *swampi.Swampi
+		endpoint2 *swampi.Swampi
+	}
```
Comment (xiegeo) [inline api/v2/api.go:69]: Why is there a hard limit on two endpoints?

**Review by xiegeo — APPROVED** [review]

File: api/v2/routes_rtfsp_management.go (hunk 1 above)
Comment (bonedaddy) [reply api/v2/routes_rtfsp_management.go:37]: yea it is

File: api/v2/api.go (hunk 2 above)
Comment (bonedaddy) [reply api/v2/api.go]: no real reason. at the moment we only need 2, however I've gone ahead and structure it in such a way that you can use more than 2 endpoints if the config package is adjusted to support it

File: api/v2/utils.go [hunk 3]
```diff
@@ -308,36 +308,33 @@ func (api *API) getCaptchaKey() string {
 // dualSwarmUpload enables uploading data to 2 different swarm nodes at once
 // to provide high-availability of stored data
 func (api *API) dualSwarmUpload(data []byte, isTar bool) (string, error) {
```
Comment (xiegeo) [inline api/v2/utils.go:310]: This should be renamed now, since the limit of 2 swarms is removed

File: api/v2/routes_swarm.go [hunk 4]
```diff
@@ -13,7 +13,7 @@ import (
 
 // SwarmUpload is used to upload data to ethereum swarm
 func (api *API) SwarmUpload(c *gin.Context) {
-	if api.swarm.endpoint1 == nil {
+	if api.swarmEndpoints == nil {
 		Fail(c, errors.New("must have at least one non nil swarm client"))
 		return
```
Comment (xiegeo) [inline api/v2/routes_swarm.go]: len(...) == 0 is more dependable than testing for nil.

Also, the error message should be updated to "must have at least one ~~non nil~~ swarm client"

File: api/v2/utils.go [hunk 5]
```diff
@@ -308,36 +308,33 @@ func (api *API) getCaptchaKey() string {
...
-		resp1.Body.Close()
-		return "", err
-	}
-	hash1 := string(responseData)
-	resp1.Body.Close()
-	if api.swarm.endpoint2 != nil {
-		resp2, err := api.swarm.endpoint2.Send(swampi.SingleFileUpload, bytes.NewReader(append(data[0:0:0], data...)), map[string][]string{
+	var hashes []string
+	for _, endpoint := range api.swarmEndpoints {
+		resp, err := endpoint.Send(swampi.SingleFileUpload, bytes.NewReader(append(data[0:0:0], data...)), map[string][]string{
```
Comment (xiegeo) [inline api/v2/utils.go:313]: I'm not sure why `append(data[0:0:0], data...` isn't just data? I think the extra data copy is unnecessary.

**Review by xiegeo — CHANGES_REQUESTED** [review]

File: api/v2/utils.go (hunk 5 above)
Comment (bonedaddy) [reply api/v2/utils.go:313]: Good catch, I thought the bytes reader would drain the data but it doesnt

File: api/v2/routes_swarm.go (hunk 4 above)
Comment (bonedaddy) [reply api/v2/routes_swarm.go]: good idea

**Review by xiegeo — APPROVED** [review]
Just a question about if `dualSwarmUpload` should be renamed left.

**Comment by bonedaddy** [comment]
@xiegeo ah shoot I forgot about that, let me make the change

**Review by xiegeo — APPROVED** [review]
